### PR TITLE
feat(MeshHTTPRoute): support hostnames with MeshGateway

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues -- #8142 
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/core/rules/merge.go [hunk 1]
```diff
@@ -10,57 +10,71 @@ import (
...
 
 	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/util/pointer"
 )
 
 // indicates that all slices that start with this prefix will be appended, not replaced
 const appendSlicesPrefix = "Append"
 
-func MergeConfs(confs []interface{}) (interface{}, error) {
+func MergeConfs(confs []interface{}) ([]interface{}, error) {
```
Comment (lobkovilya) [inline pkg/plugins/policies/core/rules/merge.go:19]: Please add a doc comment for this function, we have to explain `[]interface{}` return value

File: pkg/plugins/policies/core/rules/merge.go [hunk 2]
```diff
@@ -10,57 +10,71 @@ import (
...
-	}
 
-	confType := reflect.TypeOf(confs[0])
-	result, err := newConf(confType)
-	if err != nil {
-		return nil, err
-	}
+		if err := json.Unmarshal(resultBytes, result); err != nil {
+			return nil, err
+		}
```
Comment (lobkovilya) [inline pkg/plugins/policies/core/rules/merge.go:61]: I think this code can be moved to a separate private function:
```go
// mergeJSONPatches merges a list of confs to a single conf using the algorithm described in https://www.rfc-editor.org/rfc/rfc7396
func mergeJSONPatches(confs []interface{}) (interface{}, error) {
```
that's only for readability reasons

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go [hunk 3]
```diff
@@ -29,6 +29,10 @@ type MeshHTTPRoute struct {
 }
 
 type To struct {
+	// Hostnames limits the effects of the rules to requests to this hostname.
+	// Given hostnames must intersect with the hostname of the listeners the
```
Comment (lobkovilya) [inline pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go:33]: I'd expand the comment mentioning that by "listener" we mean both the outbound listener (with implicit "*" hostname) and meshgateway listener

**Review by lobkovilya — COMMENTED** [review]
Do we already have an e2e test for MeshHTTPRoute attached to MeshGateway?

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go (hunk 3 above)
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go:33]: clarified about MeshGateway only in the doc string

**Review by lobkovilya — APPROVED** [review]
LGTM

**Pull request merged**